streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -257,30 +257,28 @@ private long processNext(final List<ConsumerRecord<byte[], byte[]>> records,
                              final StateRestorer restorer,
                              final Long endOffset) {
         final List<KeyValue<byte[], byte[]>> restoreRecords = new ArrayList<>();
-        long nextPosition = -1;
+        long offset = -1;
 
         for (final ConsumerRecord<byte[], byte[]> record : records) {
-            final long offset = record.offset();
+            offset = record.offset();
             if (restorer.hasCompleted(offset, endOffset)) {
-                nextPosition = record.offset();
                 break;
             }
             if (record.key() != null) {
                 restoreRecords.add(KeyValue.pair(record.key(), record.value()));
             }
         }
 
-        if (nextPosition == -1) {
-            nextPosition = consumer.position(restorer.partition());
+        if (offset == -1) {
+            offset = consumer.position(restorer.partition());
         }
 
         if (!restoreRecords.isEmpty()) {
             restorer.restore(restoreRecords);
-            restorer.restoreBatchCompleted(nextPosition, records.size());
-
+            restorer.restoreBatchCompleted(offset + 1, records.size());
         }
 
-        return nextPosition;
+        return consumer.position(restorer.partition());
```

Comment:
why this? You already have the position from above

Reply:
We need to return the latest position we *fetched*. If we return `nextPostition` we return the latest position we *restored*, which might be smaller than the fetched position as we stop restoring early if we reach the initially collected endOffsets in `if (restorer.hasCompleted(offset, endOffset)) {` (L267/264). Thus the check in L240/241 (`if (pos > endOffset + 1) {`) would always be true.

It's not the best check to be honest, because as per chance a fetch might just return the "correct" amont of data such that "restore end offset" and "fetch end offset" are the same, even if there is more data available  in the changelog as another thread did take over this task and started to write to it. But I am not sure if we could fix this completely, as it seems to be a race condition anyway. Even if the restore completes perfectly fine, (in theory, we could still miss a rebalance and drop out of the group). But we would figure out in next `poll()` call; thus, it seems that this check (if the changelog end offset moved during restore), is just a "early indicator of a missed rebalance" and not strictly required.

Reply:
sounds good